In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv(r"C:\Users\jinay\Downloads\melb_data.csv")
df

Suburb           Address  Rooms Type      Price Method  \
0         Abbotsford      85 Turner St      2    h  1480000.0      S   
1         Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2         Abbotsford      5 Charles St      3    h  1465000.0     SP   
3         Abbotsford  40 Federation La      3    h   850000.0     PI   
4         Abbotsford       55a Park St      4    h  1600000.0     VB   
...              ...               ...    ...  ...        ...    ...   
13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  \
0        Biggin   3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0   
1        Biggin   4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0   
2        Biggin   4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0   
3        Biggin   4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0   
4        Nelson   4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0   
...         ...         ...       ...       ...  ...       ...  ...       ...   
13575     Barry  26/08/2017      16.7    3150.0  ...       2.0  2.0     652.0   
13576  Williams  26/08/2017       6.8    3016.0  ...       2.0  2.0     333.0   
13577     Raine  26/08/2017       6.8    3016.0  ...       2.0  4.0     436.0   
13578   Sweeney  26/08/2017       6.8    3016.0  ...       1.0  5.0     866.0   
13579   Village  26/08/2017       6.3    3013.0  ...       1.0  1.0     362.0   

       BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0               NaN        NaN        Yarra -37.79960   144.99840   
1              79.0     1900.0        Yarra -37.80790   144.99340   
2             150.0     1900.0        Yarra -37.80930   144.99440   
3               NaN        NaN        Yarra -37.79690   144.99690   
4             142.0     2014.0        Yarra -37.80720   144.99410   
...             ...        ...          ...       ...         ...   
13575           NaN     1981.0          NaN -37.90562   145.16761   
13576         133.0     1995.0          NaN -37.85927   144.87904   
13577           NaN     1997.0          NaN -37.85274   144.88738   
13578         157.0     1920.0          NaN -37.85908   144.89299   
13579         112.0     1920.0          NaN -37.81188   144.88449   

                       Regionname Propertycount  
0           Northern Metropolitan        4019.0  
1           Northern Metropolitan        4019.0  
2           Northern Metropolitan        4019.0  
3           Northern Metropolitan        4019.0  
4           Northern Metropolitan        4019.0  
...                           ...           ...  
13575  South-Eastern Metropolitan        7392.0  
13576        Western Metropolitan        6380.0  
13577        Western Metropolitan        6380.0  
13578        Western Metropolitan        6380.0  
13579        Western Metropolitan        6543.0  

[13580 rows x 21 columns]

First, I will be applying all the basic commands describe(), info() and isnull().sum() to look at the dataset further and check out if there are any missing values

In [3]:
df.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  13580.000000  1.358000e+04  13580.000000  13580.000000  13580.000000   
mean       2.937997  1.075684e+06     10.137776   3105.301915      2.914728   
std        0.955748  6.393107e+05      5.868725     90.676964      0.965921   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.500000e+05      6.100000   3044.000000      2.000000   
50%        3.000000  9.030000e+05      9.200000   3084.000000      3.000000   
75%        3.000000  1.330000e+06     13.000000   3148.000000      3.000000   
max       10.000000  9.000000e+06     48.100000   3977.000000     20.000000   

           Bathroom           Car       Landsize  BuildingArea    YearBuilt  \
count  13580.000000  13518.000000   13580.000000   7130.000000  8205.000000   
mean       1.534242      1.610075     558.416127    151.967650  1964.684217   
std        0.691712      0.962634    3990.669241    541.014538    37.273762   
min        0.000000      0.000000       0.000000      0.000000  1196.000000   
25%        1.000000      1.000000     177.000000     93.000000  1940.000000   
50%        1.000000      2.000000     440.000000    126.000000  1970.000000   
75%        2.000000      2.000000     651.000000    174.000000  1999.000000   
max        8.000000     10.000000  433014.000000  44515.000000  2018.000000   

          Lattitude    Longtitude  Propertycount  
count  13580.000000  13580.000000   13580.000000  
mean     -37.809203    144.995216    7454.417378  
std        0.079260      0.103916    4378.581772  
min      -38.182550    144.431810     249.000000  
25%      -37.856822    144.929600    4380.000000  
50%      -37.802355    145.000100    6555.000000  
75%      -37.756400    145.058305   10331.000000  
max      -37.408530    145.526350   21650.000000

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13580 non-null  object 
 1   Address        13580 non-null  object 
 2   Rooms          13580 non-null  int64  
 3   Type           13580 non-null  object 
 4   Price          13580 non-null  float64
 5   Method         13580 non-null  object 
 6   SellerG        13580 non-null  object 
 7   Date           13580 non-null  object 
 8   Distance       13580 non-null  float64
 9   Postcode       13580 non-null  float64
 10  Bedroom2       13580 non-null  float64
 11  Bathroom       13580 non-null  float64
 12  Car            13518 non-null  float64
 13  Landsize       13580 non-null  float64
 14  BuildingArea   7130 non-null   float64
 15  YearBuilt      8205 non-null   float64
 16  CouncilArea    12211 non-null  object 
 17  Lattitude      13580 non-null  float64
 18  Longti

In [5]:
print(df.dtypes)

Suburb            object
Address           object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object


In [6]:
df.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                62
Landsize            0
BuildingArea     6450
YearBuilt        5375
CouncilArea      1369
Lattitude           0
Longtitude          0
Regionname          0
Propertycount       0
dtype: int64

Firstly, I will encode all the columns that are non-numerical so that we can calculate correlation, and also because the ML model only accepts numerical data. So, encoding is necessary.

In [7]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

objects = df.select_dtypes(include = 'object').columns # Here, I am extracting those columns which are of object type.
objects 

Index(['Suburb', 'Address', 'Type', 'Method', 'SellerG', 'Date', 'CouncilArea',
       'Regionname'],
      dtype='object')

In [8]:
for i in objects:
    df[i] = le.fit_transform(df[i])

df

Suburb  Address  Rooms  Type      Price  Method  SellerG  Date  \
0           0    12794      2     0  1480000.0       1       23    45   
1           0     5943      2     0  1035000.0       1       23    47   
2           0     9814      3     0  1465000.0       3       23    48   
3           0     9004      3     0   850000.0       0       23    48   
4           0    10589      4     0  1600000.0       4      155    49   
...       ...      ...    ...   ...        ...     ...      ...   ...   
13575     302     1991      4     0  1245000.0       1       16    33   
13576     305    12234      3     0  1031000.0       3      251    33   
13577     305    12745      3     0  1170000.0       1      194    33   
13578     305    13311      4     0  2500000.0       0      222    33   
13579     313    10776      4     0  1285000.0       3      239    33   

       Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0           2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1           2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2           2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3           2.5    3067.0  ...       2.0  1.0      94.0           NaN   
4           2.5    3067.0  ...       1.0  2.0     120.0         142.0   
...         ...       ...  ...       ...  ...       ...           ...   
13575      16.7    3150.0  ...       2.0  2.0     652.0           NaN   
13576       6.8    3016.0  ...       2.0  2.0     333.0         133.0   
13577       6.8    3016.0  ...       2.0  4.0     436.0           NaN   
13578       6.8    3016.0  ...       1.0  5.0     866.0         157.0   
13579       6.3    3013.0  ...       1.0  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  Regionname  \
0            NaN           31  -37.79960   144.99840           2   
1         1900.0           31  -37.80790   144.99340           2   
2         1900.0           31  -37.80930   144.99440           2   
3            NaN           31  -37.79690   144.99690           2   
4         2014.0           31  -37.80720   144.99410           2   
...          ...          ...        ...         ...         ...   
13575     1981.0           33  -37.90562   145.16761           4   
13576     1995.0           33  -37.85927   144.87904           6   
13577     1997.0           33  -37.85274   144.88738           6   
13578     1920.0           33  -37.85908   144.89299           6   
13579     1920.0           33  -37.81188   144.88449           6   

       Propertycount  
0             4019.0  
1             4019.0  
2             4019.0  
3             4019.0  
4             4019.0  
...              ...  
13575         7392.0  
13576         6380.0  
13577         6380.0  
13578         6380.0  
13579         6543.0  

[13580 rows x 21 columns]

In [9]:
df.nunique()
# Here, I am checking how many number of unique values each column has. If the number is too high then that column is not a good feature.
# Here, columns like Address have too many unique values. So, we will drop the column.

Suburb             314
Address          13378
Rooms                9
Type                 3
Price             2204
Method               5
SellerG            268
Date                58
Distance           202
Postcode           198
Bedroom2            12
Bathroom             9
Car                 11
Landsize          1448
BuildingArea       602
YearBuilt          144
CouncilArea         34
Lattitude         6503
Longtitude        7063
Regionname           8
Propertycount      311
dtype: int64

Now, I am assuming that the price column is the one to be predicted, as in most datasets. So, I will be looking at the correlations of each column with respect to the price column. If the value of correlation is too low or too high, then it is not a good feature for ML model.

In [10]:
corr = df.corr()
print(corr['Price'])

Suburb          -0.135830
Address          0.049186
Rooms            0.496634
Type            -0.414566
Price            1.000000
Method          -0.037741
SellerG         -0.052109
Date            -0.021088
Distance        -0.162522
Postcode         0.107867
Bedroom2         0.475951
Bathroom         0.467038
Car              0.238979
Landsize         0.037507
BuildingArea     0.090981
YearBuilt       -0.323617
CouncilArea     -0.063653
Lattitude       -0.212934
Longtitude       0.203656
Regionname       0.079109
Propertycount   -0.042153
Name: Price, dtype: float64


Here, columns like address has correlation very close to 0 but very high number of unique values. So, it will be dropped. Also, columns like lattitude and longitude do not make a difference for the price column and they have high number of unique values, if we just think about it. So, they will be dropped as well. 

Some columns like Building area has very good correlation but almost 50% of it is empty. So, we will be looking at the Building area column.

In [11]:
df['BuildingArea'].describe()

count     7130.000000
mean       151.967650
std        541.014538
min          0.000000
25%         93.000000
50%        126.000000
75%        174.000000
max      44515.000000
Name: BuildingArea, dtype: float64

We will replace all the null values of building area with its mean.

In [12]:
df['BuildingArea'].fillna(value=df['BuildingArea'].mean(), inplace=True)
df['BuildingArea'].describe()

C:\Users\jinay\AppData\Local\Temp\ipykernel_6540\3334380237.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['BuildingArea'].fillna(value=df['BuildingArea'].mean(), inplace=True)


count    13580.000000
mean       151.967650
std        392.002962
min          0.000000
25%        122.000000
50%        151.967650
75%        151.967650
max      44515.000000
Name: BuildingArea, dtype: float64

In [13]:
#Looking at null value count once more
df.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                62
Landsize            0
BuildingArea        0
YearBuilt        5375
CouncilArea         0
Lattitude           0
Longtitude          0
Regionname          0
Propertycount       0
dtype: int64

Here, YearBuilt has high number of null values and also its correlation with respect to the price column is relatively low (-0.323617). So, we will also be dropping this column.

As for the null values in the Car column, they are too low of a number as compared to total number of rows (13580). So, we will simply be dropping those rows.

In [17]:
df.dropna(subset=['Car'], inplace=True) # Dropping the rows which have null values in the car column.
df.isnull().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
YearBuilt        5344
CouncilArea         0
Lattitude           0
Longtitude          0
Regionname          0
Propertycount       0
dtype: int64

Now, I will be dropping all the columns of which I have mentioned the name above.

In [20]:
df.drop(columns=['Address', 'Lattitude', 'Longtitude', 'YearBuilt'], axis=1, inplace=True)
df

Suburb  Rooms  Type      Price  Method  SellerG  Date  Distance  \
0           0      2     0  1480000.0       1       23    45       2.5   
1           0      2     0  1035000.0       1       23    47       2.5   
2           0      3     0  1465000.0       3       23    48       2.5   
3           0      3     0   850000.0       0       23    48       2.5   
4           0      4     0  1600000.0       4      155    49       2.5   
...       ...    ...   ...        ...     ...      ...   ...       ...   
13575     302      4     0  1245000.0       1       16    33      16.7   
13576     305      3     0  1031000.0       3      251    33       6.8   
13577     305      3     0  1170000.0       1      194    33       6.8   
13578     305      4     0  2500000.0       0      222    33       6.8   
13579     313      4     0  1285000.0       3      239    33       6.3   

       Postcode  Bedroom2  Bathroom  Car  Landsize  BuildingArea  CouncilArea  \
0        3067.0       2.0       1.0  1.0     202.0     151.96765           31   
1        3067.0       2.0       1.0  0.0     156.0      79.00000           31   
2        3067.0       3.0       2.0  0.0     134.0     150.00000           31   
3        3067.0       3.0       2.0  1.0      94.0     151.96765           31   
4        3067.0       3.0       1.0  2.0     120.0     142.00000           31   
...         ...       ...       ...  ...       ...           ...          ...   
13575    3150.0       4.0       2.0  2.0     652.0     151.96765           33   
13576    3016.0       3.0       2.0  2.0     333.0     133.00000           33   
13577    3016.0       3.0       2.0  4.0     436.0     151.96765           33   
13578    3016.0       4.0       1.0  5.0     866.0     157.00000           33   
13579    3013.0       4.0       1.0  1.0     362.0     112.00000           33   

       Regionname  Propertycount  
0               2         4019.0  
1               2         4019.0  
2               2         4019.0  
3               2         4019.0  
4               2         4019.0  
...           ...            ...  
13575           4         7392.0  
13576           6         6380.0  
13577           6         6380.0  
13578           6         6380.0  
13579           6         6543.0  

[13518 rows x 17 columns]

In [21]:
df.isnull().sum()

Suburb           0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
CouncilArea      0
Regionname       0
Propertycount    0
dtype: int64

In [22]:
df.duplicated().sum() # Checking for duplicated values.

np.int64(0)

Now there are no null values left and there are no duplicated rows. We can further check for outliers but as most of the columns have low number of unique values, there is no point in dropping the outlier rows as the column might be categorical in nature. We can transform the values of those columns using PowerTransformer to improve the skewness of the data.